cpp/src/arrow/types/construct.cc

```diff
@@ -71,6 +72,20 @@ Status MakeBuilder(MemoryPool* pool, const std::shared_ptr<DataType>& type,
   }
 }
 
+Status MakeStructBuilder(MemoryPool* pool, const std::shared_ptr<DataType>& type,
+    const std::vector<std::shared_ptr<Field>>& fields,
```

Comment:
Why are you defining this in terms of Fields instead of DataTypes?   This is incongruent with other builder APIs I believe.